### PR TITLE
Advanced Search updates

### LINKED ISSUE #4629: RFE: Add "movement type" filter to the 'Advanced Search' on Unit Selector
It would be useful to be able to filter vehicles by 'movement type' (i.e. wheeled/tracked/hover/etc) in the upgraded Advanced Search filter window.

### LINKED ISSUE #5076: RFE: Advanced search to filter out Unofficial Equipment
Coming from the CGL Forums

https://bg.battletech.com/forums/index.php/topic,83817.0.html

_Currently on dev release 0.49.17.  Is there a way remove -only- the Unofficial level of equipment in the Weapons/Equipment tab of Advanced Search?  I'd like these MFUK RAC/20s to disappear forever and stop cluttering up my Advanced Search filters, please.

While I'm wishing for a pony: is there a good way in the Weapons/Equipment filters to inclusively select equipment by rules categories in combination, like "All TW + All advanced" or "All TW + All advanced + All experimental"?  Because selecting "All" gives me the same MFUK RAC/20s that I don't want cluttering up the list._

### LINKED ISSUE #5096: RFE: Infantry Improvements to Advanced Search
With the recent addition of Anti-Mek gear would it be possible to add two things to the equipment search:

Ability to search for AM Gear
Ability to search for Specific Infantry armor

### LINKED ISSUE #5163: RFE: Advanced search my motive types
Was raised on another Discord.  Would it be to filter by all the supported Vehicle movement types.

### LINKED ISSUE #5715: RFE - Allow Advanced filter for maximum# of equipment (allow for excluding equipment)
### Environment
(mine) MekHQ WIndows 0.49.19.1

### Description
I assume this dialog is common to Megamek and MMLab. "Select Unit", "Advanced", "Weapons and Equipment"

You can select "has weapon type", Qty of one or more, let's say 3 of "Ultra A/C".  (As far as I know, this is not specific to Ultra AC, it's the same for everything.)

There's no way to filter for "DOES NOT HAVE an Ultra AC".  The Qty chosen seems to always mean "Show Qty or greater".

Please add ability to modify the meaning of Qty to "less than".   This way you could choose 1, apply "less than", and the result would be anything with less than 1 , i.e. zero.

image

### LINKED ISSUE #5716: RFE ~ A system of flags for "game rules concepts" and ability to filter (include/exclude) in advanced unit selection.
### Environment
(mine) MekHQ WIndows 0.49.19.1

### Description
I assume this dialog is common to Megamek and MMLab. "Select Unit", "Advanced", "Weapons and Equipment"

1. Flags would need to be included in weapons/equipment definitions, and supported in MegaMekLab.  Likely would not affect anything in gameplay.
2. Each unit design would get the logical OR of all equipment flags.
3. A new page in advanced selection might be necessary, to permit include and/or exclude of selected flags.

#### Example flags
- requires ammunition
- does damage in clusters [*]
- has a minimum range
- multiple munition types (LBX Slug/Cluster for example)
- Jump Movement
- Changes Movement (TSM, MASC)
- To-Hit adjustments (pulse lasers, heavy lasers, LBX Cluster)
- Defensive Equipment (AMS)

My thinking is generally things that "add complexity" get a flag.  This could be really useful to tailor a selection of units for newer players, to start with the simplest units possible, then add concepts gradually.

[*] I don't recall if aerospace combat does damage in clusters or not.  If not, and you're selecting aerospace units, this flag should be ignored.  Granted, at the point you're picking aerospace units, keeping things simple is a bit moot.

### PR DESCRIPTION
**Pull request by SJuliez**
This PR
- Reworks the weapons&equipment tab so the filter summary does not vanish at the bottom of the window (not part of the scroll pane); makes the weapon and equipment tables less tall
- Adds a text filter to the weapons&equipment tab that filters the tables according to the input
- reworks all scrollpanels to be less obnoxious
- Uses FlatLaf's own tristate checkbox in the unit type tab
- On the code side, takes apart TWAdvancedSearchPanel (and removes an old copy of it that wasnt used anymore)
- moves all the search stuff into its own package

image

Fixes #5715
Fixes #5163
Fixes #4629
Fixes #5076
Fixes #5096
Fixes #5716 (most of the points)

### CODE REVIEW

**Comment by HammerGS** [comment]
Have some ideas and suggestions. 

- Some equipment with variable costs shows as -2147483648 would it be possible to just show variable. (i.e AES)

- Some weapons have negative damage (i.e. Mech Mortars) since the damage is ammo based would it be possible to show variable as well.
 
- One mentioned on Discord, would be it possible to add the ability to search for Movement to the Total Warfare search versus blending with Alpha Strike. 

Noticed this. 
image

Not sure if it's us not tagging the naval weapons or a filtering issue. As the NAC's seem to filter.

**Comment by SJuliez** [comment]
Do we need the Lvl column?

**Comment by HammerGS** [comment]
> Do we need the Lvl column?

Given we have the ability in the regular selector it does duplicate. If given a choice between movement and lvl I'd vote movement.

**Comment by SJuliez** [comment]
What we have is:
image

**Comment by SJuliez** [comment]
OK - many of the filters we already have elsewhere, but the vee distinctions not so much. Still - could add those as part of the unit types instead? Note: this long list is only so I can show them all, it'll normally be in a scrollable box.
image
Add here instead?
image

**Comment by HammerGS** [comment]
That list is perfect other than the multiple UMU's. What about putting it here?

image

**Comment by SJuliez** [comment]
ok moved that list. I'm afraid writing something more useful than "UMU", "UMU" and "UMU" is going to require a significant update to those checkbox lists.

**Comment by kuronekochomusuke** [comment]
> ok moved that list. I'm afraid writing something more useful than "UMU", "UMU" and "UMU" is going to require a significant update to those checkbox lists.

can you just distinct the stream for this, since it is using the text values?

**Comment by SJuliez** [comment]
> > ok moved that list. I'm afraid writing something more useful than "UMU", "UMU" and "UMU" is going to require a significant update to those checkbox lists.
> 
> can you just distinct the stream for this, since it is using the text values?

keenly observed :)

**Comment by SJuliez** [comment]
@HammerGS I updated the source search. See tooltip

**Comment by HammerGS** [comment]
> @HammerGS I updated the source search. See tooltip

Fantastic!  All the improvements.

File: megamek/src/megamek/client/ui/advancedsearch/TriStateItem.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MegaMek. If not, see <http://www.gnu.org/licenses/>.
+ */
+package megamek.client.ui.advancedsearch;
+
+class TriStateItem {
```
[the remark this answers is not in the document]
Comment (SJuliez) [reply megamek/src/megamek/client/ui/advancedsearch/TriStateItem.java:21]: no, its text gets changed

**Comment by pavelbraginskiy** [comment]
image

Could this dropdown also include some equipment types, like if I want to search for something with any Probe or any ECM suite but don't care which one?

**Comment by kuronekochomusuke** [comment]
quirk tab is missing the clear button

**Comment by SJuliez** [comment]
> image
> 
> Could this dropdown also include some equipment types, like if I want to search for something with any Probe or any ECM suite but don't care which one?

It can contain anything. Just needs to be added

**Comment by SJuliez** [comment]
For probes, I think writing "probe" in the text filter gets them all; "ecm" will find all ECMs and a bit of other gear

**Comment by kuronekochomusuke** [comment]
can you add buttons for weapon filters similar to the ones in megameklab for the weapon tab?

image

File: megamek/src/megamek/client/ui/advancedsearch/AdvancedSearchEquipmentClass.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MegaMek. If not, see <http://www.gnu.org/licenses/>.
+ */
+package megamek.client.ui.advancedsearch;
+
+import java.util.Locale;
+
+enum AdvancedSearchEquipmentClass {
```
Comment (HammerGS) [inline megamek/src/megamek/client/ui/advancedsearch/AdvancedSearchEquipmentClass.java:23]: Could you build the list by the Flags?  For Example Watchdog doesn't have Probe or ECM but has these flags. 

or(F_ECM).or(F_BAP)

Comment (SJuliez) [reply megamek/src/megamek/client/ui/advancedsearch/AdvancedSearchEquipmentClass.java:23]: no. we are searching the cache entries, therefore we have only the names.

File: megamek/src/megamek/client/ui/advancedsearch/AdvancedSearchEquipmentClass.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            return name.contains("rotary");
+        } else if (this == ULTRA) {
+            return name.contains("ultraa");
+        } else if (this == INFANTRY_SUIT) {
+            return (name.contains("suit") || name.contains(" kit") || name.contains(", standard") || name.contains(", concealed")
+                || name.contains("clothing") || name.contains("vest") || name.contains("chainmail") || name.contains("parka"))
+                && !name.contains("ecm") && !name.contains("generic") && !name.contains("suite");
+        } else if (this == AMS) {
+            return name.contains("ams") || name.contains("antimiss");
+        } else if (this == PROBE) {
```
Comment (HammerGS) [inline megamek/src/megamek/client/ui/advancedsearch/AdvancedSearchEquipmentClass.java:65]: Should also include
Electronic Warfare (EW) Equipment
Watchdog Composite Electronic Warfare System (CEWS)
Nova Combined Electronic Warfare System (CEWS)

this is BA specific - Improved Sensors

**Comment by SJuliez** [comment]
> can you add buttons for weapon filters similar to the ones in megameklab for the weapon tab?
> 
> image

not necessary, I added the equipment group dropdown as a table filter

**Comment by SJuliez** [comment]
Going to wrap this up.

**Review by HammerGS — APPROVED** [review]
Been testing and working with this since the PR opened as draft. No issues and seems to work well.

**Review by kuronekochomusuke — APPROVED** [review]

**Pull request merged**